script/model.py

```diff
@@ -4,141 +4,350 @@
+        self.z_to_x = Decoder(2*z_size, 200, x_size)
+
     def forward(self, images):
-        self.batch_size = images.shape[0]
-        self.input = images
-        
-        self.belief_states, (h_n, c_n) = self.lstm(images)
-        self.state_mean = self.belief_to_state_mean(self.belief_states)
-        self.state_logstd = self.belief_to_state_logstd(self.belief_states)
-
-    def calculate_loss(self, t_1, t_2):
-        ## sample a state at time t_2
-        self.epsilon_t_2 = torch.randn(self.batch_size, self.state_size)
-        self.state_t_2 = self.state_mean[:,t_2,:] + \
-                         torch.exp(self.state_logstd[:,t_2,:])*self.epsilon_t_2
-
-        ## inference a state at time t_1
-        tmp = torch.cat((self.belief_states[:,t_1,:],
-                         self.belief_states[:,t_2,:],
-                         self.state_t_2), 1)
-        self.state_t_1_mean = self.infer_state_mu(tmp)
-        self.state_t_1_logstd = self.infer_state_logstd(tmp)
-        
-        self.epsilon_t_1 = torch.randn(self.batch_size, self.state_size)
-        self.state_t_1 = self.state_t_1_mean + \
-                         torch.exp(self.state_t_1_logstd)*self.epsilon_t_1
-
-        ####  calculate loss  ####
-        ## p(x_2|z_2)
-        obs_prop_t_2 = self.state_to_obs(self.state_t_2)
-        self.nloss = torch.mean(torch.sum(
-            self.input[:,t_2,:]*torch.log(obs_prop_t_2) + \
-            (1-self.input[:,t_2,:])*torch.log(1-obs_prop_t_2), 1))
-
-        ## -Pb(z_2|b_2)
-        self.nloss += -torch.mean(-torch.sum(self.epsilon_t_2**2, 1) - self.state_size/2*torch.log(2*PI) \
-                                   -torch.sum(self.state_logstd[:,t_2,:], 1))
-
-        ## Pb(z_1|b_1)
-        self.nloss += torch.mean(torch.sum(-0.5*((self.state_t_1 - self.state_mean[:,t_1,:])/torch.exp(self.state_logstd[:,t_1,:]))**2, 1) - \
-                   self.state_size/2*torch.log(2*PI) - torch.sum(self.state_logstd[:,t_1,:], 1))
-
-        ## q(z_1|z_2,b_1,b_2)
-        self.nloss += -torch.mean(-torch.sum(self.epsilon_t_1**2, 1) - self.state_size/2*torch.log(2*PI) \
-                                   -torch.sum(self.state_t_1_logstd, 1))
-
-        ## p(z_2|z_1)
-        self.state_t_2_mean = self.state_to_state_mean(self.state_t_1)
-        self.state_t_2_logstd = self.state_to_state_logstd(self.state_t_1)
-
-        self.nloss += torch.mean(torch.sum(-0.5* ((self.state_t_2 - self.state_t_2_mean)/torch.exp(self.state_t_2_logstd))**2, 1) - \
-                                 self.state_size/2*torch.log(2*PI) - torch.sum(self.state_t_2_logstd, 1))
-
-        self.loss = -self.nloss
-
-        return self.loss
-        
+        # batch size = number of rows 
+        self.batch_size = images.size()[0]
+        self.x = images
+        # pre-process image x
+        self.processed_x = self.process_x(self.x)
+
+        ####### aggregate the belief b??
+        ###### what are h_n and c_n?
```

Comment:
to clean - are `h_n` and `c_n` used?

Reply:
I am not sure. If they are it's internally by pytorch.